### PR TITLE
Placeholders in service-url for dynamic parameters (#8)

### PR DESCRIPTION
**Pull request by dprzybyl**
---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/RequestPlaceholderHelper.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+package com.cognifide.knotx.placeholder;
+
+import org.apache.commons.lang3.StringUtils;
+
+import io.vertx.core.http.HttpServerRequest;
+
+public final class RequestPlaceholderHelper {
```
Comment (Skejven) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/RequestPlaceholderHelper.java]: I'd rather see this implementation of `RequestPlaceholder`:

```java
public class RequestPlaceholder {

    private final String placeholder;

    public RequestPlaceholder(String placeholder) {
        this.placeholder = placeholder;
    }

    public String getValue(HttpServerRequest request) {
        String value = null;
        final Optional<Prefix> prefix = Prefix.of(placeholder);
        if (prefix.isPresent()) {
            value = prefix.get().getValue(request, placeholder);
        }
        return value;
    }

    private enum Prefix {
        HEADER("header.") {
            @Override
            String getValue(HttpServerRequest request, String placeholder) {
                return request.getHeader(getName(placeholder));
            }
        },
        PARAM("param.") {
            @Override
            String getValue(HttpServerRequest request, String placeholder) {
                return request.getParam(getName(placeholder));
            }
        },
        ATTRIBUTE("attribute.") {
            @Override
            String getValue(HttpServerRequest request, String placeholder) {
                return request.getFormAttribute(getName(placeholder));
            }
        };

        String prefix;

        Prefix(String prefix) {
            this.prefix = prefix;
        }

        abstract String getValue(HttpServerRequest request, String placeholder);

        boolean accept(String placeholder) {
            return StringUtils.startsWith(placeholder, prefix);
        }

        String getName(String placeholder) {
            return StringUtils.substringAfter(placeholder, prefix);
        }

        static Optional<Prefix> of(String placeholder) {
            return Arrays.stream(values()).filter(value -> value.accept(placeholder)).findFirst();
        }
    }
}
```

Comment (tomaszmichalak) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/RequestPlaceholderHelper.java]: Please implement this class as enum.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            putValues(SELECTOR, selectorString, ".", 0);
+
+            putValue(EXTENSION, matcher.group(5));
+            putValue(SUFFIX, matcher.group(6));
+        }
+    }
+
+    private void putValues(String key, String valuesString, String separator, int skip) {
+        if (valuesString != null) {
+            final int index[] = { 0 };
```
Comment (Skejven) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java]: Should this variable be an array?

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriHelper.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+package com.cognifide.knotx.placeholder;
+
+import org.jsoup.nodes.Element;
+import org.springframework.util.PropertyPlaceholderHelper;
+
+import io.vertx.core.http.HttpServerRequest;
+
+public final class UriHelper {
```
Comment (tomaszmichalak) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriHelper.java]: Please implement this class as singleton bean (not helper class), then you can write proper unit tests for this class.

http://docs.spring.io/spring/docs/current/spring-framework-reference/html/beans.html

File: knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    @Parameterized.Parameter(value = 3)
+    public String extension;
+
+    @Parameterized.Parameter(value = 4)
+    public String suffix;
+
+    @Test
+    public void test() {
+        final UriInfo uriInfo = new UriInfo(uri);
+        boolean result = StringUtils.equals(uriInfo.getValue("path"), path);
```
Comment (Skejven) [inline knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java]: Can you provide separate assertion for each check?

File: knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class UriInfoTest {
+
+    @Parameterized.Parameters(name = "{index}: {0}")
+    public static Collection<Object[]> data() {
```
Comment (tomaszmichalak) [inline knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java]: Please use zohhak library (com.googlecode.zohhak.api.TestWith).

File: knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    public String selectorString;
+
+    @Parameterized.Parameter(value = 3)
+    public String extension;
+
+    @Parameterized.Parameter(value = 4)
+    public String suffix;
+
+    @Test
+    public void test() {
```
Comment (Skejven) [inline knotx-core/src/test/java/com/cognifide/knotx/placeholder/UriInfoTest.java]: Could you provide more meaningfull test name? E.g. `testExtractingValuesFromUri` or better name following [Naming standards for unit tests](http://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html) guide.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+package com.cognifide.knotx.placeholder;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.common.collect.Maps;
+
+public class UriInfo {
```
Comment (Skejven) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java]: Could you refactor this class in similar way I proposed in `RequestPlaceholderHelper `?

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    private static final Pattern URI_PATTERN = Pattern
+            .compile("([^\\.\\?#]+)(\\.([^\\/\\?#]+)\\.)?(\\.?([^\\/\\?#]+))?([^\\?#]+)?((\\?|#).+)?");
```
Comment (tomaszmichalak) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java]: Please document from what url this pattern was copied.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriPlaceholderResolver.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.Map;
+
+import org.springframework.util.PropertyPlaceholderHelper;
+
+import com.google.common.collect.Maps;
+
+import io.vertx.core.http.HttpServerRequest;
+
+public class UriPlaceholderResolver implements PropertyPlaceholderHelper.PlaceholderResolver {
```
Comment (tomaszmichalak) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriPlaceholderResolver.java]: This class should not be spring bean in request scope.

**Comment by Skejven** [comment]
Please update documentation and examples in `README.md`.

File: knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    private static final String SELECTOR = "selector";
+
+    private static final String EXTENSION = "extension";
+
+    private static final String SUFFIX = "suffix";
+
+    private Map<String, String> values = Maps.newHashMap();
+
+    public UriInfo(String uri) {
```
Comment (tomaszmichalak) [inline knotx-core/src/main/java/com/cognifide/knotx/placeholder/UriInfo.java]: This UriInfo class should be request scope bean. Then all RequestPlaceholder's should be prototype beans with two parameters.

```java
@Bean
@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
public RequestPlaceholder requestPlaceholder(String placeholder) {
	return new RequestPlaceholder(placeholder);
}
```
Then you can use it as:
```java
applicationContext.getBean(RequestPlaceholder.class, placeholder);
```
In RequestPlaceholder class you can inject Uriinfo (which is request scope). Then in RequestPlaceholderResolver you will not keep any additional references. In that way you can simply add additional Placeholder's implementations.

**Comment by tomaszmichalak** [comment]
Add feature branch please.

**Comment by Skejven** [comment]
Closing, new pull request: https://github.com/Cognifide/knotx/pull/30

**Pull request closed without merging**